webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/RepositoryModal.js

```diff
@@ -0,0 +1,180 @@
+import LabelIcon from '../../../../../components/common/LabelIcon';
+
+import { translate as __ } from '../../../../../common/I18n';
+
+import RepositoryModalInput from './RepositoryModalInput';
+
+const RepositoryModal = ({
+  repoData,
+  handleRepoData,
+  isLoading,
+  isModalOpen,
+  handleModalToggle,
+}) => {
+  const initialRepoState = [
+    {
+      reference: 0,
+      repository: '',
+      gpgKeyUrl: '',
+    },
+  ];
+
+  useEffect(() => {
+    if (!repoData.length) {
+      handleRepoData(initialRepoState);
+    }
+  }, [repoData, initialRepoState, handleRepoData]);
+
+  const [repoReferenceCounter, setRepoReferenceCounter] = useState(0);
+
+  const updateInput = (reference, input, inputType) => {
+    handleRepoData(current =>
+      current.map(repo => {
+        if (repo.reference === reference) {
+          if (inputType === 'repository') {
+            return { ...repo, repository: input };
+          }
+
+          return { ...repo, gpgKeyUrl: input };
+        }
+        return repo;
+      })
+    );
+  };
+
+  const removeRepository = reference => {
+    handleRepoData(current =>
+      current.filter(repo => repo.reference !== reference)
+    );
+  };
+
+  const addRepositoryModalInput = () => {
+    const newRepoReferenceCounter = repoReferenceCounter + 1;
+    handleRepoData(current => [
+      ...current,
+      {
+        reference: newRepoReferenceCounter,
+        repository: '',
+        gpgKeyUrl: '',
+      },
+    ]);
+    setRepoReferenceCounter(newRepoReferenceCounter);
+  };
+
+  const clearRepositories = () => {
+    handleRepoData(initialRepoState);
+    handleModalToggle();
+  };
+
+  const renderRepo = repo => (
+    <Split hasGutter key={'repo_modal_list'.concat(repo.reference.toString())}>
+      <SplitItem span={10}>
+        <RepositoryModalInput
+          reference={repo.reference}
+          updateRepository={updateInput}
+          isLoading={isLoading}
+          initRepo={repo.repository}
+          initGpgKey={repo.gpgKeyUrl}
+        />
+      </SplitItem>
+      <SplitItem>
+        <Button
+          ouiaId={repo.reference}
+          variant="link"
+          icon=<MinusCircleIcon />
+          onClick={() => removeRepository(repo.reference)}
+        >
+          {__('Remove')}
+        </Button>
+      </SplitItem>
+    </Split>
+  );
+
+  return (
+    <Modal
+      ouiaId="host_reg_repo_modal"
+      key="host_reg_repo_modal"
+      title={__('Repository list')}
+      isOpen={isModalOpen}
+      onClose={clearRepositories}
+      actions={[
+        <Button
+          ouiaId="reg_modal_confirm"
+          key="reg_modal_confirm"
+          variant="primary"
+          onClick={handleModalToggle}
+        >
+          {__('Confirm')}
+        </Button>,
+        <Button
+          ouiaId="reg_modal_cancel"
+          key="reg_modal_cancel"
+          variant="link"
+          onClick={clearRepositories}
+        >
+          {__('Cancel')}
```

Comment:
Suppose I added 3 repositories and open the Modal again, and without editing anything I hit cancel, it removes all the repositories. Shouldn't it keep the previous state if nothing changes? Is it by design?

Reply:
I will fix this, this is not the behavior a user would expect I suppose. I could replace the `cancel` Button with a `Reset` button and make closing the modal do the same as `Confirm`. `Cancel` implies that you can discard new changes - this would make the implementation more difficult.

Reply:
Works now as expected.